File: express-backend/src/views/private.js
import express from "express";

import prisma from "../db.js";

const router = express.Router();

router.get("/:file", async (req, res) => {
    const wantedFile = req.params.file;

    const message = await prisma.message.findUnique({
        where: { file: wantedFile },
    });
    
    if (
        !message ||
        ![message.recipientId, message.senderId].includes(req.user.id)
    ) {
        return res.status(404).send("File not found");
    }

    res.status(200);
    res.header("Content-Type", "");
    res.header("X-Accel-Redirect", `/private/${wantedFile}`);
    res.end();
});

export default router;

File: django-backend/api/views/private.py
from typing import Any

from django.http import HttpResponse
from ninja import Router

from api.models import Message


router = Router(tags=["private"])


@router.get("/{file}", response={200: Any, 404: Any})
def home(request, file: str):
    if (
        msg := Message.objects.filter(file=f"private/{file}").first()
    ) is None or request.auth.id not in [msg.sender_id, msg.recipient_id]:
        return 404

    res = HttpResponse(status=200)
    res["Content-Type"] = ""
    res["X-Accel-Redirect"] = f"/private/{file}"
    return res
